src/seq_mv/csr_matvec.c

```diff
@@ -166,41 +167,201 @@ hypre_CSRMatrixMatvecOutOfPlaceHost( HYPRE_Complex    alpha,
 
       if (num_rownnz < xpar * num_rows)
       {
+         switch (num_vectors)
+         {
+            case 2:
 #ifdef HYPRE_USING_OPENMP
-         #pragma omp parallel for private(i,j,jj,m,tempx) HYPRE_SMP_SCHEDULE
+               #pragma omp parallel for private(i,j,jj,m) HYPRE_SMP_SCHEDULE
 #endif
-         for (i = 0; i < num_rownnz; i++)
-         {
-            m = A_rownnz[i];
-            for (j = 0; j < num_vectors; j++)
-            {
-               tempx = 0.0;
-               for (jj = A_i[m]; jj < A_i[m + 1]; jj++)
+               for (i = 0; i < num_rownnz; i++)
                {
-                  tempx += A_data[jj] * x_data[j * vecstride_x + A_j[jj] * idxstride_x];
+                  m = A_rownnz[i];
+
+                  HYPRE_Complex tmp[2] = {0.0, 0.0};
+                  for (jj = A_i[m]; jj < A_i[m + 1]; jj++)
+                  {
+                     HYPRE_Int     xidx = A_j[jj] * idxstride_x;
+                     HYPRE_Complex coef = A_data[jj];
+
+                     tmp[0] += coef * x_data[xidx];
+                     tmp[1] += coef * x_data[xidx + vecstride_x];
+                  }
+                  HYPRE_Int yidx = m * idxstride_y;
+
+                  y_data[yidx] += tmp[0];
+                  y_data[yidx + vecstride_y] += tmp[1];
                }
-               y_data[j * vecstride_y + m * idxstride_y] += tempx;
-            }
-         }
+               break;
+
+            case 3:
```

Comment:
Should there be a case 1 where no striding is needed?

Reply:
This situation is never reached because the switch statement is inside the `if` statement:
https://github.com/hypre-space/hypre/blob/b0c0040632b7b5795466bfa7bd4225e88c4c1aae/src/seq_mv/csr_matvec.c#L117

so decided to not include it. Let me know if you think it's better to throw a warning or something else :)

Reply:
Sorry, I missed that detail.  Looks good.